### DISCUSSION TITLE
Expose PaymentMethodIcon and types to 3rd parties

### DISCUSSION THREAD

**Discussion by itambek-godaddy** [woocommerce/woocommerce#45010] [WooCommerce, Blocks, and Site Editing]
When registering a payment method, it's possible to pass an [`icons` property](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/types/type-defs/payments.ts#L110-L111). This is not documented in the [3rd party dev docs](https://github.com/woocommerce/woocommerce-blocks/blob/21789e820e68084bc96e37405c90e7a924a26a9b/docs/third-party-developers/extensibility/checkout-payment-methods/payment-method-integration.md#props-fed-to-payment-method-nodes). Upon investigation, it looks like it affects the card logos displayed in cart blocks (example: https://share.cleanshot.com/CztY0hv4). Would it be possible to add docs around this prop so the intention is clear, and also provide a list of accepted values (card types)? Ideally, the Typescript type definitions would be exposed as well.

Related to this, a `PaymentMethodIcons` component is passed as a prop to the payment method component. As I understand, this can be used to render a list of payment method icons inside the payment method components. There are a couple of ways this component could be improved:

* It appears it requires manually passing in the icons prop again - it would be great it this was automatically done for us (if possible, while still being able to manually override), since we provide the icons property when registering the payment method.
* We’re adding card type detection to our gateways, very similar to the one in [Braintree for WooCommerce Payment Gateway](https://wordpress.org/plugins/woocommerce-gateway-paypal-powered-by-braintree/) (now managed by WooCommerce) - an example is here: https://share.cleanshot.com/3tRb00pC - however the icon inside the card number field is different from the one used by `PaymentMethodIcons` , because we don’t have access to individual [icon definitions from here](https://github.com/woocommerce/woocommerce-blocks/blob/2be0b8a00181086979eeff7d66c006108ef4afe0/assets/js/base/components/cart-checkout/payment-method-icons/common-icons.ts#L10-L106). Would it be possible to expose these definitions and utility functions, so that 3rd party gateways could load individual icons from WC blocks? Or even better, expose the `PaymentMethodIcon` component with the type definitions of each card type?.

**Comment by opr** [comment]
Hi @itambek-godaddy thanks a lot for posting this and for summarising the issues you have with the `PaymentMethodIcons` component.

> It appears it requires manually passing in the icons prop again - it would be great it this was automatically done for us (if possible, while still being able to manually override), since we provide the icons property when registering the payment method.

Hmm, I don't think it's possible to do that automatically (because the component would need context of where it's being rendered to know this) but I wonder if instead of that we can add a prop to specify which payment method it belongs to, for example:

```tsx
<PaymentMethodIcons method={ 'stripe' } />
```

Then it would read the icons specified when that method was registered. Any thoughts on this approach?

> We’re adding card type detection to our gateways, very similar to the one in [Braintree for WooCommerce Payment Gateway](https://wordpress.org/plugins/woocommerce-gateway-paypal-powered-by-braintree/) (now managed by WooCommerce) - an example is here: [share.cleanshot.com/3tRb00pC](![Dcv8HxO50SpQLJM5OkpQIZXdDH02MgWmj10SZ20G.jpg](https://github.com/woocommerce/woocommerce-blocks/assets/5656702/1030061a-a23c-492a-9e0b-898ea038f554)) - however the icon inside the card number field is different from the one used by PaymentMethodIcons , because we don’t have access to individual [icon definitions from here](https://github.com/woocommerce/woocommerce-blocks/blob/2be0b8a00181086979eeff7d66c006108ef4afe0/assets/js/base/components/cart-checkout/payment-method-icons/common-icons.ts#L10-L106). Would it be possible to expose these definitions and utility functions, so that 3rd party gateways could load individual icons from WC blocks? Or even better, expose the PaymentMethodIcon component with the type definitions of each card type?

Yes, it would be, I'd expect them to be added to the `@woocommerce/blocks-checkout`. I created https://github.com/woocommerce/woocommerce-blocks/issues/11492 to track this.

**Reply by itambek-godaddy** [reply]
> I wonder if instead of that we can add a prop to specify which payment method it belongs to

That could work - which prop would take precedence then? `icons` or `method`?

**Reply by opr** [reply]
It could use both, display the specified icons first then the ones from the method?

**Reply by itambek-godaddy** [reply]
That sounds good to me 👍
